keps/0012-framework-extensions.md

```diff
@@ -0,0 +1,319 @@
+  init:
+    resources:
+      - init.yaml
+```
+
+An extension framework, that was trying to update the `init` task with a patch captured in `templates/init-patch.yaml` could update the task in different, but equivalent ways:
+
+```yaml
+tasks:
+  init:
+    from: base/init
+    path:
+      - init-patch.yaml
+```
+
+```yaml
+tasks:
+  init:
+    resources:
+      - base/init.yaml
+    patch:
+      - init-patch.yaml
+```
+
+#### Modifying Plan
+
+When a plan is defined in an extended framework, it **replaces** the plan from the base. All plans defined in the base are available in the extension plan.
+
+#### Add and Updating Parameters
+
+Parameters can be updated in the extension framework by providing new default values or descriptions. The follow example shows how parameters can be overridden and added from the base framework.
+
+Consider the following parameter file defined in the base framework
+
+```yaml
+backup:
+  default: backup.sql
+  description: The file the backup job saves the sql dump, and the file the restore occurs from.
+password:
+  default: password
+  description: Some words
+```
+
+And this file defined in the extension framework:
+
+```yaml
+backup:
+  default: /path/to/new/location.sql
+password:
+  description: A more detailed description of the parameter
+data:
+  default: /path/to/sample/data.sql
+  description: Storage location of sample data to load
+```
+
+Would combine in the extension framework as though the following parameters file was used:
+
+```yaml
+backup:
+  description: The file the backup job saves the sql dump, and the file the restore occurs from.
+  default: /path/to/new/location.sql
+password:
+  default: password
+  description: A more detailed description of the parameter
+data:
+  default: /path/to/sample/data.sql
+  description: Storage location of sample data to load
+```
+
+#### Example Framework Extension
+
+This framework is built from the MySQL framework defined above, but adds custom plans that allow for the loading and clearing of data that is required for a particular business application, as well as a new parameter that allows for sizing the PVC that backups are stored on.
+
+```shell
+.
+├── framework.yaml
+├── params.yaml
+└── templates
+    ├── clear-data.yaml
+    ├── pvc-size-patch.yaml
+    └── load-data.yaml
+```
+
+In order to implement these changes, we need to add the plans for `load-data` and `clear-data` and update the jobs that backup and restore the data.
+
+##### framework.yaml
+
+`framework.yaml` is the base definition of a framework. It follows the following format, extracted from the MySQL example:
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+version: "5.7"
+tasks:
+  backup:
+    from: base/backup
+    patch:
+      - pvc-size-patch.yaml
+  restore:
+    from: base/restore
+    patch:
+      - pvc-size-patch.yaml
+  load-data:
+    resources:
+      - base/init.yaml
+    patches:
+      - load-data.yaml
+  clear-data:
+    from: base/init
```

Comment:
so this is the same as 

```
resources:
      - base/init.yaml
```
?

Reply:
No.  `base/init.yaml` in the resources sections refers to the element in the `templates` folder on the base.  The `base/init` in the `from` field refers to the Task `init` in the base framework.

Reply:
okay that was not clear, maybe note it somewhere